fix: make sure to catch jinja2 vars used in single-line `for` and `set` statements

beckermr opened the pull request:
<!-- Hello! Thanks for submitting a PR! To help make things go a bit more
     smoothly, we would appreciate it if you follow this template. -->

### Description

This PR adds two regexes to catch variables used in single-line jinja2 `for` and `set` statements. These have been missed for a long time and appear to be the cause of at least one bug. I have not gone through the bug tracker, but I have a hunch this PR might fix other bugs related to similar situations I've seen in the past.

closes #5445 

<!-- Good things to put here include:
       - reasons for the change (please link any relevant issues!),
       - any noteworthy (or hacky) choices to be aware of,
       - or what the problem resolved here looked like. -->

### Checklist - did you ...

<!-- If any of the following items aren't relevant to your contribution,
     please still tick them so we know you've gone through the checklist. -->

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     <!-- All "significant" changes should get an entry:
            - user-facing changes or enhancements
            - bug fixes
            - deprecations
            - documentation updates
            - other changes -->
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

<!-- Just as a reminder, everyone in all conda org spaces (including PRs)
     must follow the Conda Org Code of Conduct (link below).

     Finally, once again, thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:
       - Conda Org COC: https://github.com/conda-incubator/governance/blob/main/CODE_OF_CONDUCT.md
       - Contributing docs: ../blob/main/CONTRIBUTING.md -->

beckermr commented:
@kenodegard @isuruf or anyone else, can you review this PR? 

It is an important bug fix that I'd like to see in 24.7.2 if we can.

beckermr commented on news/5447-jinja2-for-set-vars:
```diff
@@ -0,0 +1,20 @@
+
+### Bug fixes
+
+* Variables used in jinja2 ``for`` and ``set`` statements are now properly included in the variant
```
```suggestion
* Variables used in single-line jinja2 ``for`` and ``set`` statements are now properly included in the variant
```

jaimergp commented on conda_build/variants.py:
```diff
@@ -748,12 +748,22 @@ def find_used_variables_in_text(variant, recipe_text, selectors_only=False):
         conditional_regex = (
             r"(?:^|[^\{])\{%\s*(?:el)?if\s*.*" + v_regex + r"\s*(?:[^%]*?)?%\}"
         )
+        for_regex = r"(?:^|[^\{])\{%\s*for\s*.*\s*in\s*" + v_regex + r"(?:[^%]*?)?%\}"
```
Apparently Jinja supports things like:

```jinja
{% for user in users if not user.hidden %}
    <li>{{ user.username|e }}</li>
{% endfor %}
```

or 

```jinja
<ul class="sitemap">
{%- for item in sitemap recursive %}
    <li><a href="{{ item.href|e }}">{{ item.title }}</a>
    {%- if item.children -%}
        <ul class="submenu">{{ loop(item.children) }}</ul>
    {%- endif %}</li>
{%- endfor %}
</ul>
```

, which this regex would not over I think? But  I've never seen it in the wild, so maybe it's not worth implementing here. A `TODO` comment might be in order.

jaimergp commented:
Looking at this code for the first time I wonder why are regexes needed when apparently Jinja will happily give you a list of variables in use [with three lines](https://stackoverflow.com/a/70903503). Are we doing this _before_ the selectors are parsed? Would it make sense to preprocess the selectors and then defer to the Jinja AST instead of regexes? Mentioning this because of the single-line limitations and exotic syntax not covered right now. Documenting these limitations in the function docstring might be enough.

beckermr commented:
I looked at that stackoverflow for a while, but didn't do any investigating. 

Here is how it behaves:

```python
import jinja2
import jinja2.meta

env = jinja2.Environment()
parsed_content = env.parse(
"""
{% set CL_VERSION = "19.29" %}
{% if VCVER is not defined %}
{% set BLAH = "19.29" %}
{% set FOO = "19.29" %}
{% set VCVER = "" %}
{% else %}
{% set BLAH = "19.29" %}
{% set BAZ = "19.29" %}
{% endif %}
{% set clang_major = (CLANG_VERSION|default("19.29")).split(".")[0] %}
{% set cl_minor = CL_VERSION.split(".")[1] %}
{% set vc_major = VCVER.split(".")[0] %}

package:
  name: clang-win-activation
  version: {{ CLANG_VERSION }}

build:
  number: 0
  {% if clang_major|int == 16 and cl_minor|int >= 40 %}
  skip: true
  {% endif %}

outputs:
  - name: clang_win-64
    build:
      run_exports:
        strong:
          - vc >={{ VCVER }}
    requirements:
      run:
        - clang {{ CLANG_VERSION }}.*
"""
)
tokens = jinja2.meta.find_undeclared_variables(parsed_content)

print(tokens)
```

This gives

```zsh
% python jinja.py
{'BAZ', 'VCVER', 'FOO', 'BLAH', 'CLANG_VERSION'}
```

So the behavior is basically that anything that touches a conditional with a default value or uses a default filter is marked as undefined. I'd argue that `BLAH` should not be returned since it is in both branches. 

As for when this happens, searches for jinja2 variables happens after selectors:

https://github.com/conda/conda-build/blob/d908fe1e21bcd4998bbdca02432333c9f0146b12/conda_build/metadata.py#L2856-L2892

and so using the jinja2 functions won't work unless we change the order.

beckermr commented:
Other details that are fun.

1. conda-build, depending on the options, will move text blocks around and concatenate them when searching for variables which can break jinja2 as well.
2. The conda-build docs specify that selectors are processed after jinja2 which is the opposite of the search in the sense that jinja2 vars are searched for after selectors have been applied. :/

beckermr commented on conda_build/variants.py:
```diff
@@ -745,15 +745,31 @@ def find_used_variables_in_text(variant, recipe_text, selectors_only=False):
+        # NOTE: why use a regex instead of the jinja2 parser/AST?
+        # One can ask the jinj2 parser for undefined variables, but conda-build moves whole
+        # blocks of text around when searching for variables and applies selectors to the text.
+        # So the text  reaches this function is not necessarily valid jinja2 syntax. :/
```
```suggestion
        # So the text that reaches this function is not necessarily valid jinja2 syntax. :/
```

beckermr commented on conda_build/variants.py:
```diff
@@ -745,15 +745,31 @@ def find_used_variables_in_text(variant, recipe_text, selectors_only=False):
         v_req_regex = "[-_]".join(map(re.escape, v.split("_")))
         variant_regex = rf"\{{\s*(?:pin_[a-z]+\(\s*?['\"])?{v_regex}[^'\"]*?\}}\}}"
         selector_regex = rf"^[^#\[]*?\#?\s\[[^\]]*?(?<![_\w\d]){v_regex}[=\s<>!\]]"
+        # NOTE: why use a regex instead of the jinja2 parser/AST?
+        # One can ask the jinj2 parser for undefined variables, but conda-build moves whole
+        # blocks of text around when searching for variables and applies selectors to the text.
+        # So the text that reaches this function is not necessarily valid jinja2 syntax. :/
```
@jaimergp @wolfv @h-vetinari I have added a comment explaining why we have to use a regex here. Let's merge this one instead of letting the perfect be the enemy of the good.

wolfv replied:
yeah, I think at this point it's impossible to fix this properly, sadly. LGTM!

jaimergp commented on news/5447-jinja2-for-set-vars:
```diff
@@ -0,0 +1,20 @@
+
+### Bug fixes
+
+* Variables used in single-line jinja2 ``for`` and ``set`` statements are now properly included in the variant
```
```suggestion
* Variables used in single-line jinja2 `for` and `set` statements are now properly included in the variant
```

jaimergp commented on conda_build/variants.py:
```diff
@@ -745,15 +745,31 @@ def find_used_variables_in_text(variant, recipe_text, selectors_only=False):
         variant_regex = rf"\{{\s*(?:pin_[a-z]+\(\s*?['\"])?{v_regex}[^'\"]*?\}}\}}"
         selector_regex = rf"^[^#\[]*?\#?\s\[[^\]]*?(?<![_\w\d]){v_regex}[=\s<>!\]]"
+        # NOTE: why use a regex instead of the jinja2 parser/AST?
+        # One can ask the jinj2 parser for undefined variables, but conda-build moves whole
```
```suggestion
        # One can ask the jinja2 parser for undefined variables, but conda-build moves whole
```

beckermr commented:
@kenodegard @jaimergp @jezdez It appears a dev release of conda on the canary channel is buggy and that is breaking the tests here. What do we do in these cases?

isuruf commented on conda_build/variants.py:
```diff
@@ -745,15 +745,31 @@ def find_used_variables_in_text(variant, recipe_text, selectors_only=False):
+        # TODO: this `for` regex won't catch some common cases like lists of vars, multiline
+        # jinja2 blocks, if filters on the for loop, etc.
+        for_regex = r"(?:^|[^\{])\{%\s*for\s*.*\s*in\s*" + v_regex + r"(?:[^%]*?)?%\}"
+        set_regex = r"(?:^|[^\{])\{%\s*set\s*.*\s*=\s*.*" + v_regex + r"(?:[^%]*?)?%\}"
```
This regex is not very good.
The following yaml matches `glib` and `glibc` both.
```yaml
{% set a = glibc %}
```

isuruf replied:
`set_regex = r"(?:^|[^\{])\{%\s*set\s*.*\s*=\s*.*" + v_regex + r"(?![a-zA-Z_0-9])(?:[^%]*?)?%\}"`

beckermr replied:
See https://github.com/conda/conda-build/pull/5514
